src/Interpreters/Set.cpp

```diff
@@ -543,6 +543,37 @@ BoolMask MergeTreeSetIndex::checkInRange(const std::vector<Range> & key_ranges,
     auto left_lower = std::lower_bound(indices.begin(), indices.end(), left_point, less);
     auto right_lower = std::lower_bound(indices.begin(), indices.end(), right_point, less);
 
+    /// A special case of 1-element KeyRange. It's useful for partition pruning
+    bool one_element_range = true;
+    for (size_t i = 0; i < tuple_size; ++i)
+    {
+        auto & left = left_point[i];
+        auto & right = right_point[i];
+        if (left.getType() == right.getType())
+        {
+            if (left.getType() == ValueWithInfinity::NORMAL)
+            {
+                if (0 != left.getColumnIfFinite().compareAt(0, 0, right.getColumnIfFinite(), 1))
+                {
+                    one_element_range = false;
+                    break;
+                }
+            }
```

Comment:
Should we also set `one_element_range = false` if `left.getType() != ValueWithInfinity::NORMAL` ?

Reply:
Nope. Because `left.getType() == right.getType()` means we are either both -Inf or both +Inf. This is also used in  https://github.com/ClickHouse/ClickHouse/pull/12455

Reply:
But should not we set `one_element_range = false` if both -Inf or both +Inf?

Reply:
Those ranges won't intersect with IN ordered set, thus is the same as `one_element_range = false`